File: 01-array/0107-59.py
"""
leetcode 59
"""

class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        l, r, t, b = 0, n-1, 0, n-1
        res = [[0 for _ in range(n)] for _ in range(n)]
        num, tar = 1, n*n
        while num <= tar:
            # top side: from left to right
            for i in range(l, r+1):
                res[t][i] = num
                num += 1
            t += 1
            # right side: from top to bottom
            for i in range(t, b+1):
                res[i][r] = num
                num += 1
            r -= 1
            # bottom side: from right to left
            for i in range(r, l-1, -1):
                res[b][i] = num
                num += 1
            b -= 1
            # left side: from bottom to top
            for i in range(b, t-1, -1):
                res[i][l] = num
                num += 1
            l += 1
        return res

##
# recursive
#

class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        if n == 0:
            return []
        if n == 1:
            return [[1]]
        border = [i for i in range(1, 4*(n-1)+1)] # outer 
        inner = self.generateMatrix(n-2) # inner
        res = [border[:n]] + inner;
        for i in range(1, n-1):
            t = [j+len(border) for j in res[i]]
            res[i] = [border[-1]]+t+[border[n+i-1]]
        res.append(border[2*n-2:3*n-2][::-1])
        return res

File: 01-array/0107-59.cpp
/*
 * leetcode 59
 * 
 * loop invariant:
 *
 * range [left, right), always close and open
 * top row:    from left to right
 * right col:  from top to bottom
 * bottom row: from right to left
 * left col:   from bottom to top
 *
 * then at the end, it is a special case, either one grid or more left
 *
 */

class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> res(n, vector<int>(n, 0)); 
        // this is the result, 2D vector

        int startx = 0, starty = 0; // start position for each loop
        int loop = n / 2; // how many loops?
        // if n = 3, loop = 1, the center grid has to be handle separately
        int mid = n / 2; // the center grid
        // if n = 3, the center is (1, 1)
        // if n = 5, the center is (2, 2)
        int count = 1; // the value to the matrix
        int offset = 1; // have to control the length of each side in a loop
        // after each loop, the edge shrinks one
        int i,j;
        while (loop --) {
            i = startx;
            j = starty;

            // 4 for-loop, each one side, after that it completes one loop
            // top side: [left, right)
            for (j = starty; j < n - offset; j++) {
                res[startx][j] = count++;
            }
            // right side: [top, bottom)
            for (i = startx; i < n - offset; i++) {
                res[i][j] = count++;
            }
            // bottom side: [right, left)
            for (; j > starty; j--) {
                res[i][j] = count++;
            }
            // left side: [bottom, top)
            for (; i > startx; i--) {
                res[i][j] = count++;
            }

            // to prepare for the next loop:
            // the start position has to add 1
            startx++;
            starty++;
            // the offset has to move 1
            offset += 1;
        }

        // if n is an odd number, have to deal with the center grid
        if (n % 2) {
            res[mid][mid] = count;
        }
        return res;
    }
};
